### PR TITLE
Add element ID

### LINKED ISSUE #2: Add element ID
Neo4j 5 introduces new string-based element IDs meant to replace legacy numeric IDs.

* Add `element_id()` and related methods, including a specification of expectations for their use with Neo4j ≤ 4.
* Add specification of expectations for the use of `id()` and related methods with Neo4j ≥ 5.

See also https://github.com/majensen/perlbolt/issues/50.

### PR DESCRIPTION
**Pull request by johannessen**
Initial attempt to cover the Neo4j 5 element ID (#2).

The initial proposal is to declare the behaviour of `element_id()` on Neo4j ≤ 4 to be “undefined”. This technically matches the official Neo4j documentation, which is silent on the matter, but all the official drivers actually treat it as an alias to `id()` when used with old server versions. Declaring it to be “undefined” offers maximum flexibility to driver authors here.

However, perhaps we should instead rather choose one specific behaviour in the interest of interoperability. In that case, our choice shouldn’t necessarily be the same as the official drivers’ choice. Yielding the legacy numeric ID on a call to `element_id()` seems pretty useless – nonsensical, even. You can’t use it with Cypher’s `elementId()` function. That’s actually a bit of a problem for us because the element ID may well be unavailable to Perl drivers even with Neo4j 5 because of network protocol issues (e. g. Jolt v1). The official first-party drivers don’t have that problem.

Therefore, perhaps `element_id()` should provide something like `undef` instead when that value is unavailable.

A similar question arises about the future behaviour of `id()`, once the legacy numeric ID becomes unavailable in Neo4j (maybe in version 6). The initial proposal doesn’t define this either, but perhaps it should try to do so.

Not yet included here is a deprecation warning for `id()`. Such a warning should be allowed with Neo4j 5+, but not with earlier versions, because it doesn’t make sense for users of old server versions to use anything other than `id()`.

### CODE REVIEW

**Comment by johannessen** [comment]
The longer I think about this, the more I’m leaning towards requiring the `element_id()` methods to return `undef` for an unavailable element ID. I believe that in most cases, it’ll make things easier for implementors and users alike.

**Comment by johannessen** [comment]
Which behaviour would be the most sensible one really depends on what you use `element_id()` for.

* If you use it to compare e. g. two nodes for equality from within Perl, returning the numeric ID makes this really easy. You can then use a comparison like `$n1->element_id eq $n2->element_id` and it’ll just work fine on every Neo4j version.

* If you use it together with Cypher’s `elementId()` function, there is no way it can possibly work. Ideally, for this scenario drivers would act fail-fast and just `die` with an appropriate error message.

Returning `undef` actually is rather problematic in both of these cases:

`undef` evaluates to an empty string, so comparison of two such values within Perl will always yield a truthy result. So now suddenly *all* your nodes might erroneously be considered equal to each other. This has the potential for data corruption. Yes, Perl will give you a warning, but it might already be too late by the time you see it. You do have good backups *(right?)* – but still, this very much seems like an unnecessary risk.

And for Cypher’s `elementId()` function, you probably won’t even be getting a warning for `undef`, because the driver will happily accept it as a parameter value, converting it to a Cypher Null value. Null will of course never match any Neo4j element ID, which means your app will just stop working right, without any indication of why.

The latter is incidentally the same situation we currently have with `element_id()` returning the numeric ID. So switching to `undef` actually wouldn’t really help with that in practical usage.

On the other hand, having `element_id()` just `die` makes writing Perl code that can compare nodes for equality only slightly more cumbersome, if that code is supposed to work on any Neo4j version. You can just use `eval` and fail over to numeric IDs. Of course, code that is intended to work only with Neo4j ≤ 4, or only with Neo4j ≥ 5, can simply use the appropriate method and doesn’t need to jump through these hoops.

```perl
use v5.16;  # eval return values were unreliable on v5.14 and earlier

my $equal = eval { $n1->element_id eq $n2->element_id } // $n1->id == $n2->id;
```

- - -

Bottom line: Requiring `element_id()` to `die` noisily on an unavailable element ID looks pretty attractive.

**Comment by johannessen** [comment]
The thing I want to try and avoid most of all here are bugs slipping past unnoticed which involve the use of Cypher’s `elementId()` function with something that’s not in fact an element ID.

Having our `element_id()` methods just `die` on Neo4j ≤ 4 achieves this. But so does a warning. For example like this:

```perl
package Neo4j::Bolt::Node {
  sub element_id ($self) {
    if ( ! defined $self->{element_id} ) {
      warnings::warnif 'Neo4j::Types', 'Element ID is unavailable';
      return $self->{id};  # Fall back to legacy numeric ID
    }
    return $self->{element_id};
  }
}

...

# On Neo4j ≤ 4:
@nodes = $stream->fetch_next;

$equal = $nodes[0]->element_id eq $nodes[1]->element_id;
# Actually compares legacy numeric IDs, not element IDs.
# Issues warning "Element ID is unavailable", but otherwise works as expected.

{
  no warnings 'Neo4j::Types';
  $element_id = $nodes[2]->element_id;
}
$cxn->run_query(q{MATCH (a) WHERE elementId(a) = $eid RETURN a}, {eid => $element_id});
# Compares the element IDs of nodes with a legacy numeric ID parameter.
# Always fails on Neo4j ≤ 4 without a warning. I can still shoot myself in the foot
# if I absolutely have to, but it’s a bit harder to do here than it was before.
```

Just like using `eval`, having a `no warnings` block is also a bit cumbersome. But with a requirement that a warning “must be issued by default”, implementors could conceivably choose to offer other, less annoying ways to suppress that warning.

**Comment by johannessen** [comment]
I think #13 is the way to go. Closing this.
